orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -47,53 +52,92 @@ Status Optimizer::ConstructInputs() {
+                                               .GetMutable<Tensor>();
+        second_order_moments.emplace_back(
+            Tensor(second_order_moment_tensor->DataType(), second_order_moment_tensor->Shape(),
+                   second_order_moment_tensor->MutableDataRaw(), second_order_moment_tensor->Location()));
       } else {
-        ORT_ENFORCE("This is an invalid graph. Optimizer graph contains unknown user input:", name);
+        ORT_ENFORCE(
+            false, "This is an invalid graph. Optimizer graph contains unknown user input:", input_names_[i]);
       }
     }
+
+    const auto tensorseq_inserter = [](auto& tensors, auto* inputs) {
+      ORT_ENFORCE(!tensors.empty(), "Tensors cannot be empty while building a tensor sequence.");
+
+      auto tensor_seq = std::make_unique<TensorSeq>(tensors.front().DataType());
+      tensor_seq->SetElements(std::move(tensors));
+      inputs->emplace_back(
+          OrtValue(tensor_seq.release(), DataTypeImpl::GetType<TensorSeq>(),
+                   DataTypeImpl::GetType<TensorSeq>()->GetDeleteFunc()));
+    };
+
+    // Add the params and moments as tensorseq ortvalues to inputs
+    tensorseq_inserter(params, &inputs_);
+    tensorseq_inserter(first_order_moments, &inputs_);
+    tensorseq_inserter(second_order_moments, &inputs_);
+
+    // Add the gradients as ortvalues to inputs
+    inputs_.insert(inputs_.end(),
+                   std::make_move_iterator(grads.begin()),
+                   std::make_move_iterator(grads.end()));
   }
   // Add other optimizer reordering logic here
   return Status::OK();
 }
 
 Optimizer::Optimizer(const std::string& optim_path_or_bytes,
-                     const std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters) : named_parameters_(parameters) {
+                     const std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters)
+    : named_parameters_(parameters) {
   // TODO: share threadpool, env with module session
   const SessionOptions session_options;
   std::unique_ptr<Environment> env;
   ORT_ENFORCE(Environment::Create(nullptr, env) == Status::OK(), "Enviroment creation fails.");
   optim_sess_ = std::move(std::make_unique<InferenceSession>(session_options, *env));
-
   ORT_THROW_IF_ERROR(optim_sess_->Load(optim_path_or_bytes));
   ORT_THROW_IF_ERROR(optim_sess_->Initialize());
 
   utils::GetGraphInputOutputNames(optim_sess_, input_names_, output_names_);
-  ORT_ENFORCE(input_names_[0] == "learning_rate");  // TODO: make this better
-  ORT_ENFORCE(input_names_[1] == "step");           // TODO: make this better
+  ORT_ENFORCE(input_names_[0] == LearningRateName);        // TODO: make this better
+  ORT_ENFORCE(input_names_[1] == StepName);                // TODO: make this better
+  ORT_ENFORCE(input_names_[2] == ParamsName);              // TODO: make this better
+  ORT_ENFORCE(input_names_[3] == FirstOrderMomentsName);   // TODO: make this better
```

Comment:
we need a decoupling on this kind of hardcoding in the Optimizer Step based on the optimizer type. We shouldn't assume that we have moments as inputs.

Reply:
moved it inside the if.